packages/state-transition/src/cache/epochCache.ts

```diff
@@ -714,9 +753,33 @@ export class EpochCache {
     return isAggregatorFromCommitteeLength(committee.length, slotSignature);
   }
 
+  getPubkey(index: ValidatorIndex): PublicKey {
+    return this.globalIndex2pubkey[index];
+  }
+
+  getValidatorIndex(pubkey: Uint8Array | PubkeyHex): ValidatorIndex | undefined {
+    return this.globalPubkey2index.get(pubkey) || this.unfinalizedPubkey2index.get(toMemoryEfficientHexStr(pubkey));
+  }
+
+  /**
+   *
+   * Add unfinalized pubkeys
+   *
+   */
   addPubkey(index: ValidatorIndex, pubkey: Uint8Array): void {
-    this.pubkey2index.set(pubkey, index);
-    this.index2pubkey[index] = bls.PublicKey.fromBytes(pubkey, CoordType.jacobian); // Optimize for aggregation
+    this.unfinalizedPubkey2index.set(toMemoryEfficientHexStr(pubkey), index);
```

Comment:
This entire feature should be fork gated. If this feature in merged to unstable (best option) it must be developed in a way that before eip6110 activation there is 0 performance penalty both in compute and memory footprint. The simplest path would be to assume that before eip6110 activation all pubkeys being added to this function are finalized

Reply:
That makes perfect sense! Will address it